### DISCUSSION TITLE
Discussion on Dataflow IR

### DISCUSSION THREAD

**Discussion by KuangjuX** [TiledTensor/TiledCUDA#134] [General]
Discussion on the Definition of IR Data Structure Between Parser and CodeEmitter.

**Comment by KuangjuX** [comment]
I would like to discuss the data structures and internal definitions used in data flow analysis, including the following parts:
- Dataflow Graph
- Dataflow Node
    - Buffer Node
    - Operator Node
    - Block Node
- Dataflow Edge
    - Common Edge
    - Attached Edge
- Access Map

**Reply by haruhi55** [reply]
Take a [simple example of GEMM](https://github.com/TiledTensor/TiledCUDA/blob/master/examples/cpp/gemm/gemm.cu#L7-L30). In an ideal situation, the nodes and edges in the graph directly correspond to the syntactic elements in TiledCuda.

<p align="center">
<img src="https://github.com/user-attachments/assets/2d158e69-2615-41e9-b1f3-9fc9861f5649" width=50%>
</p>

**Comment by KuangjuX** [comment]
`BufferNode` represents a block of memory and requires specifying `shape`, `typing`, `name`, `memory_level` and `id`:

```rust
pub struct Buffer {
    name: String,
    id: usize,
    shape: Shape,
    typing: Typing,
    memory_level: MemoryLevel
}
```
